synchronization/futex.rs

```diff
@@ -0,0 +1,524 @@
+#![feature(core_intrinsics)]
+
+use libc::{syscall, timespec, SYS_futex, SYS_gettid};
+use nix::errno::Errno;
+use nix::{errno, Error as NixError, Result};
+use std::time;
+
+use lock_api::RawMutex;
+
+use std::sync::atomic::{AtomicI32, Ordering};
+
+use std::cell::RefCell;
+use std::intrinsics::{likely, unlikely};
+use std::thread_local;
+use unreachable::unreachable;
+
+const FUTEX_TID_MASK: libc::pid_t = 0x3fffffff;
+
+/// Number of tasks to wake when waking using a futex.
+pub enum WakeNumber {
+    /// Wake a certain number of tasks
+    N(u32),
+    /// Wake all tasks (platform-defined number)
+    All,
+}
+
+/// Trait representing the basic primitives used in this library for futexes.
+/// Higher level functionality is implemented on top of this.
+pub trait RawFutex {
+    const INIT: Self;
+
+    /// Wakes a `WakeNumber` of tasks waiting on this futex.
+    fn wake(&self, wake: WakeNumber) -> Result<i32>;
+
+    /// Wait for the futex to be set. If it is set, will return immediately.
+    /// On error will return the errno value.
+    fn wait(&self, value: i32, timeout: Option<time::Duration>) -> Result<i32>;
+
+    /// This wraps the syscall used after the user-space lock attempt fails. It checks if the
+    /// futex is nonzero, and if so sets the `FUTEX_WAITERS` bit, and then attaches the waiter
+    /// futex. The waiting occurs in priority order.
+    fn lock_pi(&self, value: i32, timeout: Option<time::Duration>) -> Result<i32>;
+
+    /// Attempts to acquire the lock. This is more robust than simply checking the atomic because
+    /// the kernel might be able to acquire the lock in cases where the user code lacks the necessary
+    /// information (i.e if user-space things are stale or someone died).
+    fn trylock_pi(&self, value: i32, timeout: Option<time::Duration>) -> Result<i32>;
+
+    /// Wakes the top priority waiter.
+    fn unlock_pi(&self) -> Result<i32>;
+
+    /// User-space atomic operation allowing a compare and swap operation on the user-space atomic
+    /// data in the futex.
+    fn compare_and_swap(&self, current_value: i32, new_value: i32, ordering: Ordering) -> i32;
```

Comment:
What makes this necessary, given that the implementation is just an inline wrapper for a c&s call with essentially the same semantics?

Reply:
so that the impl for RawFutex is comprehensive.

Since we need to do a c&s operation on the futex's underlying data, a MockFutex needs to expose the same interface.